store/cache/cache.go

```diff
@@ -0,0 +1,112 @@
+)
+
+var (
+	_ types.CommitKVStore             = (*CommitKVStoreCache)(nil)
+	_ types.MultiStorePersistentCache = (*CommitKVStoreCacheManager)(nil)
+
+	// DefaultCommitKVStoreCacheSize defines the persistent ARC cache size for a
+	// CommitKVStoreCache.
+	DefaultCommitKVStoreCacheSize uint = 1000
+)
+
+type (
+	// CommitKVStoreCache implements an inter-block (persistent) cache that wraps a
+	// CommitKVStore. Reads first hit the internal ARC (Adaptive Replacement Cache).
+	// During a cache miss, the read is delegated to the underlying CommitKVStore
+	// and cached. Deletes and writes always happen to both the cache and the
+	// CommitKVStore in a write-through manner. Caching performed in the
+	// CommitKVStore and below is completely irrelevant to this layer.
+	CommitKVStoreCache struct {
+		types.CommitKVStore
+		cache *lru.ARCCache
+	}
+
+	// CommitKVStoreCacheManager maintains a mapping from a StoreKey to a
+	// CommitKVStoreCache. Each CommitKVStore, per StoreKey, is meant to be used
+	// in an inter-block (persistent) manner and typically provided by a
+	// CommitMultiStore.
+	CommitKVStoreCacheManager struct {
+		cacheSize uint
+		caches    map[string]types.CommitKVStore
+	}
+)
+
+func NewCommitKVStoreCache(store types.CommitKVStore, size uint) *CommitKVStoreCache {
+	cache, err := lru.NewARC(int(size))
+	if err != nil {
+		panic(fmt.Errorf("failed to create KVStore cache: %s", err))
+	}
+
+	return &CommitKVStoreCache{
+		CommitKVStore: store,
+		cache:         cache,
+	}
+}
+
+func NewCommitKVStoreCacheManager(size uint) *CommitKVStoreCacheManager {
+	return &CommitKVStoreCacheManager{
+		cacheSize: size,
+		caches:    make(map[string]types.CommitKVStore),
+	}
+}
+
+// GetStoreCache returns a Cache from the CommitStoreCacheManager for a given
+// StoreKey. If no Cache exists for the StoreKey, then one is created and set.
+// The returned Cache is meant to be used in a persistent manner.
+func (cmgr *CommitKVStoreCacheManager) GetStoreCache(key types.StoreKey, store types.CommitKVStore) types.CommitKVStore {
+	if cmgr.caches[key.Name()] == nil {
+		cmgr.caches[key.Name()] = NewCommitKVStoreCache(store, cmgr.cacheSize)
+	}
+
+	return cmgr.caches[key.Name()]
+}
+
+// Reset resets in the internal caches.
+func (cmgr *CommitKVStoreCacheManager) Reset() {
+	cmgr.caches = make(map[string]types.CommitKVStore)
+}
+
+// CacheWrap returns the inter-block cache as a cache-wrapped CommitKVStore.
+func (ckv *CommitKVStoreCache) CacheWrap() types.CacheWrap {
+	return cachekv.NewStore(ckv)
+}
+
+// Get retrieves a value by key. It will first look in the write-through cache.
+// If the value doesn't exist in the write-through cache, the query is delegated
+// to the underlying CommitKVStore.
+func (ckv *CommitKVStoreCache) Get(key []byte) []byte {
+	valueI, ok := ckv.cache.Get(string(key))
+	if ok {
+		// cache hit
+		return valueI.([]byte)
+	}
+
+	// cache miss; write to cache
+	value := ckv.CommitKVStore.Get(key)
+	ckv.cache.Add(string(key), value)
```

Comment:
Should we be checking if `value != nil` first before adding?

Reply:
The underlying `CommitKVStore` is guaranteed never to have a `nil` value. But i've added checks just to be safe.